### PR TITLE
Save one copy of BlockLocation

### PR DESCRIPTION
**Pull request by jiacheliu3**
In `DefaultBlockMaster` we create extra copies of `BlockLocation` while we can just use the one given.

### CODE REVIEW

File: core/server/master/src/main/java/alluxio/master/block/DefaultBlockMaster.java [hunk 1]
```diff
@@ -995,12 +995,7 @@ private void processWorkerAddedBlocks(MasterWorkerInfo workerInfo,
           Optional<BlockMeta> block = mBlockStore.getBlock(blockId);
           if (block.isPresent()) {
             workerInfo.addBlock(blockId);
-            BlockLocation blockLocation = BlockLocation.newBuilder()
-                .setWorkerId(workerInfo.getId())
-                .setTier(entry.getKey().getTier())
-                .setMediumType(entry.getKey().getMediumType())
-                .build();
-            mBlockStore.addLocation(blockId, blockLocation);
+            mBlockStore.addLocation(blockId, entry.getKey());
```
Comment (jiacheliu3) [inline core/server/master/src/main/java/alluxio/master/block/DefaultBlockMaster.java:998]: Now that the `BlockLocation` contains the worker ID, I don't see a reason why we don't use that.

File: core/server/master/src/test/java/alluxio/master/block/BlockMasterTest.java [hunk 2]
```diff
@@ -76,8 +76,6 @@
   private static final Map<Block.BlockLocation, List<Long>> NO_BLOCKS_ON_LOCATION
       = ImmutableMap.of();
   private static final Map<String, StorageList> NO_LOST_STORAGE = ImmutableMap.of();
-  private static final Block.BlockLocation BLOCK_LOCATION = Block.BlockLocation.newBuilder()
-      .setTier(Constants.MEDIUM_MEM).setMediumType(Constants.MEDIUM_MEM).build();
```
Comment (jiacheliu3) [inline core/server/master/src/test/java/alluxio/master/block/BlockMasterTest.java:80]: The manually simulated `BlockLocation` objects now should contain the correct worker IDs. There can be more tests that I haven't found, that need to be updated.

File: core/server/master/src/main/java/alluxio/master/block/BlockMasterWorkerServiceHandler.java [hunk 3]
```diff
@@ -81,7 +81,8 @@ public void blockHeartbeat(BlockHeartbeatPRequest request,
             .stream()
             .collect(
                 Collectors.toMap(
-                    e -> Block.BlockLocation.newBuilder().setTier(e.getKey().getTierAlias())
+                    e -> Block.BlockLocation.newBuilder().setWorkerId(workerId)
+                        .setTier(e.getKey().getTierAlias())
```
Comment (jiacheliu3) [inline core/server/master/src/main/java/alluxio/master/block/BlockMasterWorkerServiceHandler.java:85]: We have the worker ID here already, so we can just set it to the `BlockLocation` and pass it to the `processWorkerAddedBlocks()` method. The worst case is, the blocks are unknown and we just ignore them.

I don't really see a reason why we ignore the worker ID here. The worker IDs don't change anywhere in the `register()` or `heartbeat()`. We just look for the `MasterWorkerInfo` with the ID and then get the ID from the `MasterWorkerInfo` again... Another reason might be we don't add the worker ID here so the hash code doesn't contain worker ID, so the `BlockLocation` items are together with other `BlockLocation` from other workers in some map. But that also doesn't make sense...

File: core/server/master/src/main/java/alluxio/master/block/DefaultBlockMaster.java (hunk 1 above)
Comment (ggezer) [reply core/server/master/src/main/java/alluxio/master/block/DefaultBlockMaster.java:998]: This cut seems promising!
Can you also assert that workerIds match as expected?

**Review by ggezer — CHANGES_REQUESTED** [review]

File: core/server/master/src/main/java/alluxio/master/block/DefaultBlockMaster.java [hunk 4]
```diff
@@ -995,7 +995,14 @@ private void processWorkerAddedBlocks(MasterWorkerInfo workerInfo,
           Optional<BlockMeta> block = mBlockStore.getBlock(blockId);
           if (block.isPresent()) {
             workerInfo.addBlock(blockId);
-            mBlockStore.addLocation(blockId, entry.getKey());
+            BlockLocation location = entry.getKey();
```
Comment (ggezer) [inline core/server/master/src/main/java/alluxio/master/block/DefaultBlockMaster.java:998]: nit: Preconditions.checkState()

**Review by ggezer — APPROVED** [review]

**Comment by jiacheliu3** [comment]
rebased to resolve conflicts with the other PR https://github.com/Alluxio/alluxio/pull/13390 Now this is ready to merge.

**Comment by jiacheliu3** [comment]
alluxio-bot, merge this please

**Pull request merged**